code/modules/crafting/table.dm

```diff
@@ -20,18 +20,18 @@
 			return 0
 	return 1
 
-/obj/structure/table/proc/check_table()
+/obj/structure/table/proc/check_table(datum/table_recipe/R)
 	table_contents = list()
 	for(var/obj/item/I in loc)
 		if(istype(I, /obj/item/stack))
 			var/obj/item/stack/S = I
 			table_contents[I.type] += S.amount
 		else
-			if(istype(I, /obj/item/weapon/reagent_containers))
-				for(var/datum/reagent/R in I.reagents.reagent_list)
-					table_contents[R.type] += R.volume
-
 			table_contents[I.type] += 1
+	for(var/obj/item/weapon/reagent_containers/RC in loc)
+		if((RC.flags & OPENCONTAINER) || (R && (RC.type in R.reqs)))
+			for(var/datum/reagent/A in RC.reagents.reagent_list)
+				table_contents[A.type] += A.volume
```

Comment:
All for(each) loops will loop through everything in the location, they will just typecheck to kick out unfitting shit and typecast explicitly the fitting ones, so second loop is pretty much never a good idea

Problem with special cases of containers is that tablecrafting absolutely lacks the idea of discrimination of items based on anything thus trying to make it have it with few lines is just a hack